File: OtrisymNMF/OtrisymNMF_CD.py
import numpy as np
import time
import math
from scipy.sparse import find, csr_matrix
from .SVCA import SSPA
from .SVCA import SVCA
from .Utils import orthNNLS
from scipy.sparse import issparse
from scipy.sparse import diags


def OtrisymNMF_CD(X, r, numTrials=1, maxiter=1000, delta=1e-7, time_limit=300, init_method=None,
                  verbosity=1, init_seed=None):
    """
    Orthogonal Symmetric Nonnegative Matrix Trifactorization using Coordinate Descent.
    Given a symmetric matrix X >= 0, finds matrices W >= 0 and S >= 0 such that X ≈ WSW' with W'W=I.
    W is represented by:
    - v: indices of the nonzero columns of W for each row.
    - w: values of the nonzero elements in each row of W.

    Application to community detection:
        - X is the adjacency matrix of an undirected graph.
        - OtrisymNMF detects r communities.
        - v assigns each node to a community.
        - w indicates the importance of a node within its community.
        - S describes interactions between the r communities.

    Parameters:
        X : crs_matrix , shape (n, n)
            Symmetric nonnegative matrix (Adjacency matrix of an undirected graph).
        r : int
            Number of columns of W (Number of communities).
        numTrials : int, default=1
            Number of trials with different initializations.
        maxiter : int, default=1000
            Maximum iterations for each trial.
        delta : float, default=1e-7
            Convergence tolerance.
        time_limit : int, default=300
            Time limit in seconds.
        init_method : str, default=None
            Initialization method ("random", "SSPA", "SVCA", "SPA").
        verbosity : int, default=1
            Verbosity level (1 for messages, 0 for silent mode).
        init_seed : float, optional (default=None)
            Random seed for the initialization for the experiments

    Returns:
        w_best : np.array, shape (n,)
            Values of the nonzero elements for each row.
        v_best : np.array, shape (n,)
            Indices of the nonzero columns of W.
        S_best : np.array, shape (r, r)
            Central matrix.
        error_best : float
            Relative error ||X - WSW'||_F / ||X||_F.
        time_per_iteration : list[list[float]]
            Time per iteration for each trial.
            Each inner list contains the times (in seconds) for every iteration
            of that specific trial.
    """
    start_time = time.time()
    if not issparse(X):
        X = csr_matrix(X)
    n = X.shape[0]
    error_best = float('inf')
    time_per_iteration = []

    # Precomputations

    normX = np.sqrt(np.sum(X.data ** 2))
    I, J, V = find(X)
    perm = np.argsort(I)
    I = I[perm]
    J = J[perm]
    V = V[perm]
    rowStart = np.concatenate((
        [0],
        np.where(np.diff(I) != 0)[0] + 1,
        [len(I)]
    ))
    diagX = X.diagonal()

    if verbosity > 0:
        print(f'Running {numTrials} Trials in Series')

    for trial in range(numTrials):
        time_per_iteration.append([])
        if init_method is None:
            init_algo = "SSPA" if trial == 0 else "SVCA"
        else:
            init_algo = init_method

        # Initialization
        if init_algo == "random":
            if init_seed is not None:
                init_seed += 10 * trial
                np.random.seed(init_seed)
            base = np.arange(r)
            rest = np.random.randint(0, r, size=n - r)
            v = np.concatenate([base, rest])
            np.random.shuffle(v)
            w = np.random.rand(n)


        else:
            if init_seed is not None:
                init_seed += 10 * trial
            W = initialize_W(X, r, method=init_algo, init_seed=init_seed)
            w, v = extract_w_v(W, r)

        # Normalization of w
        nw = np.bincount(v, weights=w ** 2, minlength=r)
        nw = np.sqrt(nw)
        w = np.divide(w, nw[v], out=np.zeros_like(w), where=nw[v] != 0)

        # Compute of S
        prodVal = w[I] * w[J] * V  # w_i * w_j * X(i,j)
        S = np.zeros((r, r))
        np.add.at(S, (v[I], v[J]), prodVal)
        dgS = np.diag(S).copy()

        # Iterative update
        prev_error = compute_error(normX, S)
        error = prev_error



        for iteration in range(maxiter):
            start_it = time.time()
            if time.time() - start_time > time_limit:
                print('Time limit passed')
                break

            # Update W

            # Pre-calculations to avoid a double loop on ‘n’
            wp2 = np.zeros(r)
            S2 = S ** 2
            w2 = w ** 2
            for k in range(r):
                wp2[k] = np.sum(w2 * S2[v, k])
            # for each node
            for i in range(n):
                if time.time() - start_time > time_limit:
                    print('Time limit passed')
                    break
                # b coefficients for the r problems ax^4+bx^2+cx
                b = 2 * (wp2 - (w[i] * S[v[i], :]) ** 2) - 2 * diagX[i] * dgS

                # c coefficients
                ind = np.arange(rowStart[i], rowStart[i + 1])
                mask = J[ind] != i
                cols_i = J[ind[mask]]
                xip = V[ind[mask]]
                c = -4 * np.dot(xip * w[cols_i], S[v[cols_i], :])

                vi_new, wi_new, f_new = -1, -1, np.inf
                for k in range(r): # Test each community

                    # Cardan resolution for min ax^4+bx^2+cx
                    roots = cardan_depressed(4 * S2[k, k], 2 * b[k], c[k])

                    # best positive solution
                    x = np.sqrt(r / n)  # default value
                    min_value = S2[k, k] * (x ** 4) + b[k] * (x ** 2) + c[k] * x
                    for sol in roots:
                        value = S2[k, k] * (sol ** 4) + b[k] * (sol ** 2) + c[k] * sol
                        if sol > 0 and value < min_value:
                            x, min_value = sol, value

                    if S2[k, k] * x ** 4 + b[k] * x ** 2 + c[k] * x < f_new:
                        f_new, wi_new, vi_new = S2[k, k] * x ** 4 + b[k] * x ** 2 + c[k] * x, x, k

                # update wp2

                wp2 = wp2 - (w[i] * S[v[i], :]) ** 2 + (wi_new * S[vi_new, :]) ** 2

                # update of w et v for node i
                w[i], v[i] = wi_new, vi_new

            # Normalization of w
            nw = np.bincount(v, weights=w ** 2, minlength=r)
            nw = np.sqrt(nw)
            w = np.divide(w, nw[v], out=np.zeros_like(w), where=nw[v] != 0)

            # Compute of S
            prodVal = w[I] * w[J] * V  # w_i * w_j * X(i,j)
            S = np.zeros((r, r))
            np.add.at(S, (v[I], v[J]), prodVal)
            dgS = np.diag(S).copy()

            prev_error = error
            error = compute_error(normX, S)

            time_per_iteration[-1].append(round(time.time() - start_it, 4))

            if error < delta or abs(prev_error - error) < delta:
                break

        if iteration == maxiter - 1:
            print('Not converged')

        if error <= error_best:
            w_best, v_best, S_best, error_best = w, v, S, error
        if verbosity > 0:
            print(
                f'Trial {trial + 1}/{numTrials} with {init_algo} in {iteration} iterations: Error {error:.4e} | Best: {error_best:.4e}')

        if error_best <= delta or time.time() - start_time > time_limit:
            break

    return w_best, v_best, S_best, error_best, time_per_iteration


def initialize_W(X, r, method="SVCA", init_seed=None):
    """ Initializes W based on the chosen method."""

    if method == "SSPA":
        n = X.shape[0]
        p = max(2, math.floor(0.1 * n / r))

        options = {'average': 1}
        if init_seed is not None:
            np.random.seed(init_seed)

        if issparse(X):
            # Estimation of WS=W*S with SSPA
            # Attention the rank of X must be greater than r
            WS, K = SSPA(X.tocsc(), r, p, options=options)
            norm2x_squared = X.multiply(X).sum(axis=0)  # matrice 1 x n (sparse)
            norm2x_squared = np.array(norm2x_squared).ravel()
            norm2x = np.sqrt(norm2x_squared)
            norm2x_safe = norm2x + 1e-16

            # Normalization of the columns of X
            inv_norms = 1.0 / norm2x_safe
            D = diags(inv_norms)
            Xn = X @ D
            # Solve ||X-WSHO||_F with HOHO^T = D
            HO = orthNNLS(X, WS, Xn)
            # Transposition of the result
            W = HO.T
        else:
            # Estimation of W with SSPA
            # Attention the rank of X must be greater than r
            WS, K = SSPA(X, r, p, options=options)

            # Normalization of the X columns
            norm2x = np.sqrt(np.sum(X ** 2, axis=0))
            Xn = X * (1 / (norm2x + 1e-16))
            # Solve ||X-WSHO||_F with HOHO^T = D
            HO = orthNNLS(X, WS, Xn)
            # Transposition
            W = HO.T

    if method == "SVCA":
        n = X.shape[0]
        p = max(2, math.floor(0.1 * n / r))

        options = {'average': 1}
        if init_seed is not None:
            np.random.seed(init_seed)

        if issparse(X):
            # Estimation of WS=W*S with SVCA
            # Attention the rank of X must be greater than r
            WS, K = SVCA(X.tocsc(), r, p, options=options)
            norm2x_squared = X.multiply(X).sum(axis=0)  # matrice 1 x n (sparse)
            norm2x_squared = np.array(norm2x_squared).ravel()
            norm2x = np.sqrt(norm2x_squared)
            norm2x_safe = norm2x + 1e-16

            # Normalization of the X columns
            inv_norms = 1.0 / norm2x_safe
            D = diags(inv_norms)
            Xn = X @ D
            # Solve ||X-WSHO||_F with HOHO^T = D
            HO = orthNNLS(X, WS, Xn)
            # Transposition
            W = HO.T


        else:
            # Estimation of WS=W*S with SVCA
            # Attention the rank of X must be greater than r
            WS, K = SVCA(X, r, p, options=options)

            # Normalization of the X columns
            norm2x = np.sqrt(np.sum(X ** 2, axis=0))
            Xn = X * (1 / (norm2x + 1e-16))
            # Solve ||X-WSHO||_F with HOHO^T = D
            HO = orthNNLS(X, WS, Xn)
            # Transposition
            W = HO.T

    return W


def extract_w_v(W, r):
    """ Extracts w and v from W."""
    w = np.max(W, axis=1)
    v = np.argmax(W, axis=1)
    # assign random communities if no assignement
    zero_indices = np.where(w == 0)[0]
    random_values = np.random.randint(0, r, size=zero_indices.shape[0])
    v[zero_indices] = random_values
    return w, v


def cardan_depressed(a, c, d, tol=1e-12):
    """ Cardano formula to find the roots of ax^3+cx+d=0 """
    if abs(a) < tol:
        if abs(c) < tol:
            return []
        else:
            return [-d / c]

    # b=0 t^3+pt+q
    p = c / a
    q = d / a
    Delta = 4 * (p ** 3) + 27 * (q ** 2)

    if abs(Delta) < tol:
        return [0]
    elif Delta > 0:  # one real solution
        sqrtD = np.sqrt(Delta / 27)
        return [np.cbrt((-q + sqrtD) / 2) + np.cbrt((-q - sqrtD) / 2)]

    else:  # 3 real different solutions or multiple solution

        r = 2 * np.sqrt(-p / 3)
        cos_arg = -q / 2 * np.sqrt(-27 / (p ** 3))
        cos_arg = np.clip(cos_arg, -1, 1)
        theta = np.arccos(cos_arg) / 3
        return r * np.cos(np.array([theta, theta + 2 * np.pi / 3, theta + 4 * np.pi / 3]))


def compute_error(normX, S):
    """ Computes error ||X - WSW'||_F / ||X||_F."""
    error = np.sqrt(normX ** 2 - np.linalg.norm(S, 'fro') ** 2) / normX
    return error


def Community_detection_SVCA(X, r, numTrials=1, verbosity=1, init_seed=None):
    """
        Perform community detection using the SVCA (Smooth VCA).

        Parameters:
        - X: ndarray or sparse matrix csr_matrix (Adjacency matrix of the graph)
        - r: int (Number of communities)
        - numTrials: int (Number of trials to find the best decomposition, default=1)
        - verbosity: int (Level of verbosity for printing progress, default=1)

        Returns:
        - w_best: ndarray (Importance of each node in its community)
        - v_best: ndarray (Community index for each node)
        - S_best: ndarray (Interaction matrix between communities)
        - error_best: float (Reconstruction error of the best trial)
        """

    I, J, V = find(X)
    perm = np.argsort(I)
    I = I[perm]
    J = J[perm]
    V = V[perm]
    error_best = float('inf')
    normX = np.sqrt(np.sum(X.data ** 2))
    if verbosity > 0:
        print(f'Running {numTrials} Trials in Series')

    for trial in range(numTrials):

        if init_seed is not None:
            init_seed += 10 * trial
        W = initialize_W(X, r, method="SVCA", init_seed=init_seed)
        w, v = extract_w_v(W, r)
        # Normalization of w
        nw = np.bincount(v, weights=w ** 2, minlength=r)
        nw = np.sqrt(nw)
        w = np.divide(w, nw[v], out=np.zeros_like(w), where=nw[v] != 0)

        # Compute of S
        prodVal = w[I] * w[J] * V  # w_i * w_j * X(i,j)
        S = np.zeros((r, r))
        np.add.at(S, (v[I], v[J]), prodVal)


        error = compute_error(normX, S)

        if error <= error_best:
            w_best, v_best, S_best, error_best = w, v, S, error

        if verbosity > 0:
            print(f'Trial {trial + 1}/{numTrials} with SVCA: Error {error:.4e} | Best: {error_best:.4e}')

    return w_best, v_best, S_best, error_best

File: matlab/algo/OtrisymNMF/OtrisymNMF_CD.m
% Orthogonal Symmetric Nonnegative Matrix Trifactorization (OtrisymNMF)
% with a block coordinate descent approach
%
% function [w,v,S,error] = OtrisymNMF_CD(X,r,varargin)
%
% Heuristic to solve the following problem:
%   Given a symmetric nonnegative matrix X>=0, find matrices W>=0 and S>=0 
%   such that XWSW' with W'W=I,
%
% INPUTS
%   X: Symmetric nonnegative matrix (adjacency matrix of an undirected graph)
%   r: Number of columns of W (number of communities)
%
% Options (varargin)
%   numTrials: Number of trials with different initializations
%              Default: 1
%   init: Initialization method for W
%         Options: 'random', 'SSPA', 'SVCA' (default), 'SPA'
%   maxiter: Maximum number of iterations per trial
%            Default: 1000
%   delta: Convergence tolerance
%          Default: 1e-7
%          (Stop if the change in error between iterations is < delta or if error < delta)
%   time_limit: Time limit for a trial in seconds
%               Default: 60*5
%   verbosity: Display messages (1) or not (0)
%              Default: 1

%
% OUTPUTS
%
%   w_best: Vector of length n; w(i) gives the value of the non-zero element
%           in the i-th row of W
%   v_best: Vector of length n; v(i) gives the index of the column of W
%           corresponding to the non-zero element in the i-th row
%   S_best: Central matrix of size r x r 
%   error_best: Relative error ||X-WSW||_F/||X||_F
%   time_global: Total Runtime
%   time_iteration: time_iteration{t}{i} time of iteration i in trial t
%
% This code is a supplementary material to the paper
%  TOCOMPLETE 

function [w_best,v_best,S_best,error_best,time_global,time_iteration] = OtrisymNMF_CD(X,r,varargin)

if nargin <= 2
    options = [];
else
    for k = 1:2:length(varargin)
        options.(varargin{k}) = varargin{k+1};
    end
end
% Default Value 
if ~isfield(options, 'numTrials')
    options.numTrials = 1;
end
if ~isfield(options, 'time_limit')
    options.time_limit = 5*60;
end 
if ~isfield(options, 'delta')
    options.delta = 1e-5;
end 
if ~isfield(options, 'maxiter')
    options.maxiter = 1000;
end 
if ~isfield(options, 'verbosity')
    options.verbosity = 1;
end 
time_iteration = {};
start = tic;
error_best = 'inf';

if any(sum(X, 2) == 0)
    error(['The sparse matrix contains at least one zero row. ' ...
           'Please remove empty rows (nodes without connections) during preprocessing.']);
end

 if options.verbosity > 0
        fprintf('Running %u Trials in Series \n', options.numTrials);
 end
 
% PRECOMPUTATION 
n = size(X,1);
normX = norm(X, 'fro');
normX2 = normX^2;
[I,J,V] = find(X);    
[~, perm] = sort(I, 'ascend'); % permutation to sort given W rows
I = I(perm);
J = J(perm);
V = V(perm);
rowStart = [1; find(diff(I))+1; numel(I)+1];
 
 
for trials = 1:options.numTrials
    
    %INITIALISATION
    time = {};
    w = zeros(n,1);
    v = zeros(n,1);
    if isfield(options, 'init')
        init_algo = options.init;
    else
        init_algo = "SVCA";
    end 
    if init_algo == "random"
        for i = 1:n
            v(i) = randi([1,r]);
            w(i) = rand;
        end 
        
    elseif init_algo == "SSPA"
        % Estimation of WO=WS by SSPA
        p = max(2,floor(0.1*n/r));
        options1.average = 1;
        [WO,~] = SSPA(X,r,p,options1);
        % Compute W s.t. min ||X-WO*W'||_F W'W=I
        norm2x = sqrt(sum(X.^2, 1));
        Xn = X .* (1 ./ (norm2x + 1e-16));
        HO = orthNNLS(X, WO, Xn);
        W = HO';
        % Compute v and w given W
        v = max ((W ~= 0).*(1:r),[],2);
        zero_idx = find(v == 0);
        v(zero_idx) = randi(r, size(zero_idx)); 
        w = W(sub2ind(size(W), (1:n)', v));
        
    elseif init_algo == "SVCA"
        % Estimation of WO=WS by SVCA
        p = max(2, floor(0.1*n/r));
        options1.average = 1;
        [WO,~] = SVCA(X, r, p, options1);
        % Compute W s.t. min ||X-WO*W'||_F W'W=I
        norm2x = sqrt(sum(X.^2, 1));
        Xn = X .* (1 ./ (norm2x + 1e-16));
        HO = orthNNLS(X, WO, Xn);
        W = HO';
        % Compute v and w given W
        v = max((W ~= 0).*(1:r),[],2);
        zero_idx = find(v == 0);
        v(zero_idx) = randi(r, size(zero_idx)); 
        w = W(sub2ind(size(W), (1:n)', v));
                
    elseif init_algo == "SPA"
        % Estimation of WO=WS by SPA
        p = 1;
        options1.average = 1;
        [WO,~] = SSPA(X, r, p, options1);
        % Compute W s.t. min ||X-WO*W'||_F W'W=I
        norm2x = sqrt(sum(X.^2, 1));
        Xn = X .* (1 ./ (norm2x + 1e-16));
        HO = orthNNLS(X, WO, Xn);
        W = HO';
        % Compute v and w given W
        v = max((W~=0).*(1:r),[],2);
        zero_idx = find(v == 0);
        v(zero_idx) = randi(r, size(zero_idx)); 
        w = W(sub2ind(size(W), (1:n)', v));
    end
    
    % Normalization of w 
    colNorm = sqrt(accumarray(v,w.^2,[r 1]));
    w       = w./(colNorm(v)+1e-10);
    
    % Construction of S
    prodVal = w(I).*w(J).*V; %calcul des termes w_i*w_j*X(i,j)
    S       = accumarray([v(I),v(J)],prodVal,[r,r]);
    dgS     = diag(S)';

    % UPDATE error 
    error_pre = sqrt(normX2-norm(S,'fro')^2)/normX;
    error = error_pre;
    
    for itt = 1:options.maxiter
        start_it = tic;
        
        if toc(start) > options.time_limit
            disp('Time limit passed');
            break;
        end
        
        % Precomputation
        p  = zeros(1,r);
        S2 = S.^2;
        for k = 1:r
            p(k) = sum(w.^2 .* S2(v,k));
        end

        % UPDATE W
        for i = 1 : n
            % b coefficients of the r problems  min_x ax^4+bx^2+cx
            b = 2*(p-(w(i)*S(v(i),:)).^2)-2*X(i,i)*dgS; % O(r)

            % c coefficients of the r problems  min_x ax^4+bx^2+cx
            ind    = rowStart(i):rowStart(i+1)-1; % indices nonzeros X(i,:) 
            mask   = J(ind) ~= i;                 % remove i index
            cols_i = J(ind(mask));                % indices nonzeros de X(i,:) without i !
            xip    = V(ind(mask));                %  X values for non zeros entries indices X(i,:) without i !
            c      = -4 * ( (xip(:).*w(cols_i))' * S( v(cols_i) , : ) );   % O( nnz dans X(i,:) )

            % Solve r problems  min_x ax^4+bx^2+cx
            best_f = inf; best_x =sqrt(r/n) ; best_k = 1;
            for k = 1:r
                [x,f] = cardan_depressed(4*S2(k,k),2*b(k),c(k),sqrt(r/n));
                   
                if f < best_f
                    best_f = f; best_x = x; best_k = k;
                end
            end

            % Update of p before updating w(i) (O(r))
            p = p - (w(i)*S(v(i),:)).^2 + (best_x*S(best_k,:)).^2;

            % Update w(i)
            w(i) = best_x;
            v(i) = best_k;
        end

        % Normalization of w O(n)
        colNorm = sqrt(accumarray(v,w.^2,[r 1]));
        w       = w./(colNorm(v)+1e-10);

        % Construction of S O(nnz(X)) S = W^TXW when W^TW=I
        prodVal = w(I).*w(J).*V; % calcul w_i*w_j*X(i,j) O(nnz(X))
        S       = accumarray([v(I),v(J)],prodVal,[r,r]); % O(nnz(X) + r^2)
        dgS     = diag(S)';
    
        error_pre = error;
        error = sqrt(normX2-norm(S,'fro')^2)/normX;
        time{end+1} = toc(start_it);
        if error < options.delta
            break;
        end
        if abs(error_pre-error) < options.delta
            break;
        end
    end

    time_iteration{end+1} = time;
    time_global = toc(start);

    if error<=error_best
        w_best = w;
        v_best = v;
        S_best = S;
        error_best = error;
        if error_best <= options.delta
            break;
        end 
         if toc(start) > options.time_limit
            if options.verbosity > 0
                fprintf('Time_limit reached \n')
            end 
            break;
        end
    end 
    if options.verbosity > 0
        if itt == options.maxiter
                fprintf('Not converged \n')
        end 
        fprintf('Trial %u of %u with %s : %2.4e | Best: %2.4e \n',...
            trials, options.numTrials, init_algo, error, error_best);
            
    end
    
end 

